src/libre/dialect/sql/parser.sid

```diff
@@ -75,260 +63,265 @@
+  		};
 
-		char: () -> (c :char) = {
-			(c, !, !) = CHAR;
-		};
-
-		class: () -> (k :fsm) = {
-			k = CLASS_alnum;
-		||
-			k = CLASS_alpha;
+		atom-suffix: () -> (f :ast_count) = {
+			STAR;
+			f = <atom-kleene>();
 		||
-			k = CLASS_digit;
+			PLUS;
+			f = <atom-plus>();
 		||
-			k = CLASS_lower;
+			OPT;
+			f = <atom-opt>();
 		||
-			k = CLASS_space;
+			(pos_of, pos_ot) = OPENCOUNT;
+			m = COUNT;
+			(pos_cf, pos_ct) = CLOSECOUNT;
+			f = <expr-count>(m, pos_of, m, pos_ct);
 		||
-			k = CLASS_spchr;
+			(pos_of, pos_ot) = OPENCOUNT;
+			m = COUNT;
+			SEP;
+			n = COUNT;
+			(pos_cf, pos_ct) = CLOSECOUNT;
+			f = <expr-count>(m, pos_of, n, pos_ct);
 		||
-			k = CLASS_upper;
+			f = <atom-one>();
 		};
 
-		range: () -> (a :char, b :char) = {
-			{
-				(a, start, !) = CHAR;
-			};
-
-			{
-				(!, !, !) = RANGE;
-			##
-				<err-expected-range>;
-			};
-
-			{
-				(b, !, end) = CHAR;
-			||
-				(b, !, end) = RANGE;
-			};
-
-			<mark-range>(start, end);
-		};
-
-		list-of-terms: (g :grp &) -> () = {
-			{
-				c = char;
-				<group-add-char>(&g, c);
-			||
-				k = class;
-				<group-add-class>(&g, k);
-			||
-				(a, b) = range;
-				<group-add-range>(&g, a, b);
-			##
-				<err-expected-term>;
-			};
-
-			{
-				$;
+	        /* [abc] -> when it reaches the end, make an empty class, and
+		 *     each character literal / named class / etc. modifies it.
+		 *     
+		 * in the case of the SQL dialect's wonky [abc^def], it should
+		 * create two groups, and then remove anything from the former
+		 * that is in the latter.
+		 *
+		 * [abc] literals
+		 * [a-c] range
+		 * [-^] characters only allowed at the head
+		 * [-] also allowed at the tail (just [-] is ambiguous)
+		 * [[:NAME:]] named character classes
+		 *
+		 * _ is the ANY character class
+		 * */
+		char-class: () -> (node :ast_expr) [
+			char-class-head: () -> (f :char_class_ast) = {
+				(!) = INVERT;
+				f = <char-class-ast-flag-invert>();
 			||
-				list-of-terms(&g);
-			};
-		};
-
-		group-body: (g :grp &) -> () = {
-			{
-				(c, !, !) = CLOSEGROUP;
-				<group-add-char>(&g, c);
-			||
-				(c, !, !) = RANGE;
-				<group-add-char>(&g, c);
-			||
-				/*
-				 * XXX: this is atypical, but serves as a workaround for below
-				 * until this can be done lexically by /-$/ -> $char.
-				 */
-				(a, !, !) = CLOSEGROUP;
-				<group-add-char>(&g, a);
-				(b, !, !) = RANGE;
-				<group-add-char>(&g, b);
+				(!, !, !) = RANGE;
+				f = <char-class-ast-flag-minus>();
 			||
-				$;
-			};
-
-			list-of-terms(&g);
-
-			{
-/* XXX: this ought to be permitted, but I can't see how for LL(1)
-				(c, !, !) = RANGE;
-				<group-add-char>(&g, c);
+				(!) = INVERT;
+				(!, !, !) = RANGE;
+				f = <char-class-ast-flag-invert-minus>();
 			||
-*/
-				$;
+				f = <char-class-ast-flag-none>();
 			};
-		};
 
-		group-bm: (g :grp &) -> () = {
-			(start, !) = OPENGROUP;
+			char-class-tail: () -> (f :char_class_ast) = {
+			/* This requires LL(2). */
```

Comment:
SID does actually have a mechanism for something like this - you can supply a predicate function, written in C, which can run whatever you like. The result influences which alt to take. It's used in TenDRA for parts of C++'s grammar which are not resolvable without consulting a symbol table, or whatever else. (And also for the typedef problem in C, I think). It really suits ad-hoc lookahead. Now's not the time for it though.